src/blog/jspi.md

```diff
@@ -0,0 +1,396 @@
+
+We are definitely looking forward to seeing what you can do with this new capability!
+
+
+## Appendix A: Complete Listing of `badfib`
+
+
+```
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+#include <emscripten.h>
+
+typedef long (testFun)(long,int);
+
+#define microSeconds (1000000)
+
+long add(long x, long y){
+ return x+y;
+}
+
+// Ask JS to do the addition
+EM_JS(long, jsAdd, (long x, long y), {
+ return x+y;
+});
+
+// promise an addition
+EM_ASYNC_JS(long, promiseAdd, (long x,long y), {
+ return Promise.resolve((k,r)=>{setTimeout(()=> {return k(x+y)},0)});
+}
+);
+
+__attribute__((noinline))
+long localFib(long x){
+ if(x==0)
+   return 0;
+ if(x==1)
+   return 1;
+ return add(localFib(x-1),localFib(x-2));
+}
+
+__attribute__((noinline))
+long jsFib(long x){
+ if(x==0)
+   return 0;
+ if(x==1)
+   return 1;
+ return jsAdd(jsFib(x-1),jsFib(x-2));
+}
+
+__attribute__((noinline))
+long promiseFib(long x){
+ if(x==0)
+   return 0;
+ if(x==1)
+   return 1;
+ return promiseAdd(promiseFib(x-1),promiseFib(x-2));
+}
+
+long runLocal(long x,int count){
+ long temp = 0;
+ for(int ix=0;ix<count;ix++)
+   temp += localFib(x);
+ return temp/count;
+}
+
+long runJs(long x,int count){
+ long temp = 0;
+ for(int ix=0;ix<count;ix++)
+   temp += jsFib(x);
+ return temp/count;
+}
+
+long runPromise(long x,int count){
+ long temp = 0;
+ for(int ix=0;ix<count;ix++)
+   temp += promiseFib(x);
+ return temp/count;
+}
+
+double runTest(testFun test,int limit,int count){
+ clock_t start = clock();
+ test(limit,count);
+ clock_t stop = clock();
+ return ((double)(stop-start))/CLOCKS_PER_SEC;
+}
+
+void runTestSequence(int step,int limit,int count){
+ for(int ix=0;ix<=limit;ix+=step){
+   double light = (runTest(runLocal,ix,count)/count)*microSeconds;
+   double jsTime = (runTest(runJs,ix,count)/count)*microSeconds;
+   double promiseTime = (runTest(runPromise,ix,count)/count)*microSeconds;
+   printf("fib(%d) %gμs %gμs %gμs %gμs\n",ix, light, jsTime, promiseTime, (promiseTime-jsTime));
+ }
+}
+
+EMSCRIPTEN_KEEPALIVE int main(){
+ int step =  1;
+ int limit = 15;
+ int count = 1000;
+ runTestSequence(step,limit,count);
+ return 0;
+}
+```
+
+## Appendix B: Listing of `u42.c` and `p42.c`
+
+The `u42.c` C code represents the main part of our dynamic loading example:
+
+```
+#include <stdio.h>
+#include <emscripten.h>
+
+typedef long (*fooFun)();
+
+// promise a function
+EM_ASYNC_JS(fooFun, resolveFun, (), {
+ console.log("loading promise42");
+ LoadedModule = (await WebAssembly.instantiateStreaming(fetch("p42.wasm"))).instance;
+ return addFunction(LoadedModule.exports['provide42']);
+}
+);
+
+EM_JS_DEPS(funDeps,"$addFunction")
+
+extern fooFun get42;
+
+long stub(){
+ get42 = resolveFun();
+ return get42();
+}
+
+fooFun get42 = stub;
+
+int main(){
+ printf("first call p42() = %ld\n",get42());
+ printf("second call = %ld\n",get42());
+}
+```
+
+The p42.c code is the dynamically loaded module.
+
+```
```

Comment:
To enable syntax-highlighting, specify the programming language following the opening three backticks. E.g. in this case:

```suggestion
```c
```

Consider doing the same for other code snippets throughout the article.

Reply:
done